hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncTableRegionLocatorImpl.java

```diff
@@ -62,7 +62,10 @@ public CompletableFuture<List<HRegionLocation>> getAllRegionLocations() {
           .thenApply(locs -> Arrays.asList(locs.getRegionLocations()));
       }
       return ClientMetaTableAccessor
-        .getTableHRegionLocations(conn.getTable(TableName.META_TABLE_NAME), tableName);
+        .getTableHRegionLocations(conn.getTable(TableName.META_TABLE_NAME), tableName)
+        .whenComplete((locs, error) -> {
```

Comment:
You need to handle the case where error != null here. In that case I believe locs will be null, and you'll get an NPE below.  Can you add a test for this?

Also, I think the convention in the async client is to use FutureUtils.addListener. The benefit there is it will also handle catching any errors thrown by your own callback.

Reply:
Thanks for the congrats!

Here's the javadoc for whenComplete:

> Returns a new CompletionStage with the same result or exception as this stage, that executes the given action when this stage completes.
> When this stage is complete, the given action is invoked with the result (or null if none) and the exception (or null if none) of this stage as arguments. The returned stage is completed when the action returns.
Unlike method handle, this method is not designed to translate completion outcomes, so the supplied action should not throw an exception. However, if it does, the following rules apply: if this stage completed normally but the supplied action throws an exception, then the returned stage completes exceptionally with the supplied action's exception. Or, if this stage completed exceptionally and the supplied action throws an exception, then the returned stage completes exceptionally with this stage's exception.

My understanding of that is, if there were an error the callback would get called with `locs == null` and `error != null`. And then, your callback should not throw an exception. But since you're callback is called with locs == null, your callback is-was would throw an NPE. Then basd on the last 2 sentences, it would be the same result for the end-user, but they'd see an NPE instead of the originating error.

Reply:
Thanks for your patient reply. It helps me a lot.

After re-reading the doc and doing some tests, it turns out I was wrong before. Yes, there will indeed be NPE problem here. But based on the doc above and my tests, the user will still see the originating error. If the error is not null here, the future will completed exceptionally with this error no matter the supplied action throws an exception or not. Am I right?

Reply:
No problem, thank you for verifying in tests. I think now that you are using addListener, you are correct. This is part of the benefit of addListener, but still better to avoid the exception altogether. I believe in your original implementation with your own `withComplete` call it would have resulted in an NPE to the user (but I could be wrong, it's been a while).

Reply:
Would you mind including a test case in TestAsyncNonMetaRegionLocator for the exceptional case?

Reply:
OK. I'll address it as soon as possible. Thanks.